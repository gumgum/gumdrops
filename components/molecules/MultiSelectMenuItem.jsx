import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from './Checkbox';
import cx from 'classnames';

const baseClass = 'gds-multi-select__menu-link';

const MultiSelectMenuItem = ({
    callback,
    children,
    index,
    isActive,
    name,
    onChange,
    onClick,
    size,
    selected,
    subOptions,
    value
}) => (
    <li className="gds-multi-select__menu-item" onClick={onClick}>
        <div
            className={cx(baseClass, {
                [`${baseClass}--collapsable`]: children,
                [`${baseClass}--active`]: isActive
            })}>
            <div
                className="gds-form-group gds-multi-select__option"
                onClick={event => event.stopPropagation()}>
                <Checkbox
                    label={name}
                    checked={selected}
                    size={size}
                    onChange={() => {
                        callback && callback(index, value, !selected);
                        onChange &&
                            onChange({ index, value, selected: !selected, options: subOptions });
                    }}
                />
            </div>
            {children}
        </div>
    </li>
);

MultiSelectMenuItem.propTypes = {
    name: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    isActive: PropTypes.bool,
    size: PropTypes.string,
    value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
    selected: PropTypes.bool,
    children: PropTypes.node,
    callback: PropTypes.func,
    onChange: PropTypes.func,
    onClick: PropTypes.func,
    subOptions: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
            selected: PropTypes.bool
        })
    )
};

export default MultiSelectMenuItem;
