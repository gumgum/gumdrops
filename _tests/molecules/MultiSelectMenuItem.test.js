/**
 * @jest-environment jsdom
 */
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import MultiSelectMenuItem from '../../components/molecules/MultiSelectMenuItem';

const subOptions = [
    { name: 'Lassie', value: 1, selected: false },
    { name: 'Snoopy', value: 2, selected: false },
    { name: 'Toto', value: 3, selected: false },
    { name: 'Brian Griffin', value: 3, selected: false }
];

const defaultProps = {
    callback: jest.fn(),
    index: 1,
    isActive: false,
    name: 'Foo',
    onChange: jest.fn(),
    onClick: jest.fn(),
    selected: false,
    size: 'xs',
    value: 0
};

test('Expect <MultiSelectMenuItem> to render properly', () => {
    const tree = renderer.create(<MultiSelectMenuItem {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Expect <MultiSelectMenuItem> to render properly', () => {
    const tree = renderer.create(<MultiSelectMenuItem {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Expect <MultiSelectMenuItem> to call the onChange prop handler when options change', () => {
    const { getByLabelText } = render(<MultiSelectMenuItem {...defaultProps} />);
    fireEvent.click(getByLabelText(defaultProps.name));
    expect(defaultProps.onChange).toHaveBeenCalledWith({
        index: defaultProps.index,
        value: defaultProps.value,
        selected: !defaultProps.selected,
        options: defaultProps.options
    });
});

test('Expect <MultiSelectMenuItem> to call the onChange prop handler with sub options', () => {
    const props = {
        ...defaultProps,
        subOptions,
        onChange: jest.fn(),
        callback: jest.fn()
    };
    const { getByLabelText } = render(<MultiSelectMenuItem {...props} />);
    fireEvent.click(getByLabelText(defaultProps.name));
    expect(props.onChange).toHaveBeenCalledWith({
        index: props.index,
        value: props.value,
        selected: true,
        options: props.subOptions
    });
});
