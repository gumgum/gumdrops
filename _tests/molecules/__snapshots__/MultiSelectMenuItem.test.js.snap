// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <MultiSelectMenuItem> to render properly 1`] = `
<li
  className="gds-multi-select__menu-item"
  onClick={[MockFunction]}
>
  <div
    className="gds-multi-select__menu-link"
  >
    <div
      className="gds-form-group gds-multi-select__option"
      onClick={[Function]}
    >
      <div
        className="gds-form-group__checkbox gds-form-group__checkbox--xs"
      >
        <label
          className="gds-form-group__checkbox-label"
        >
          <input
            checked={false}
            className="gds-form-group__checkbox-input"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="gds-form-group__checkbox-indicator"
          />
          Foo
        </label>
      </div>
    </div>
  </div>
</li>
`;

exports[`Expect <MultiSelectMenuItem> to render properly 2`] = `
<li
  className="gds-multi-select__menu-item"
  onClick={[MockFunction]}
>
  <div
    className="gds-multi-select__menu-link"
  >
    <div
      className="gds-form-group gds-multi-select__option"
      onClick={[Function]}
    >
      <div
        className="gds-form-group__checkbox gds-form-group__checkbox--xs"
      >
        <label
          className="gds-form-group__checkbox-label"
        >
          <input
            checked={false}
            className="gds-form-group__checkbox-input"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="gds-form-group__checkbox-indicator"
          />
          Foo
        </label>
      </div>
    </div>
  </div>
</li>
`;
